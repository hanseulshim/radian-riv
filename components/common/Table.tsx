import React, { useEffect } from 'react'
import { useSortBy, useTable, useFlexLayout } from 'react-table'

export default function Table({ columns, data, fetchData, sortTable = false }) {
  const { getTableProps, headerGroups, rows, prepareRow } = useTable(
    {
      columns,
      data,
      disableSortRemove: true,
      initialState: {
        sortBy: [
          columns.map(one => {
            return {
              desc: true,
              id: one.accessor ? one.accessor : ''
            }
          })
        ]
      }
    },
    useSortBy,
    useFlexLayout
  )

  useEffect(() => {
    fetchData()
  }, [fetchData])

  const headerProps = (props, { column }) =>
    getStyles(props, column.align, column)

  const cellProps = (props, { cell }) => getStyles(props, cell.column.align)

  const getStyles = (props, align = 'left', column?) => {
    if (!column || !sortTable) {
      return [
        props,
        {
          style: {
            justifyContent: align === 'right' ? 'flex-end' : 'flex-start'
          }
        }
      ]
    }
    const colProps = column.getSortByToggleProps()
    return [
      props,
      {
        ...colProps,
        style: {
          ...colProps.style,
          justifyContent: align === 'right' ? 'flex-end' : 'flex-start'
        }
      }
    ]
  }

  return (
    <div className="table-container">
      <div {...getTableProps()} className="table">
        <div className="thead">
          {headerGroups.map(headerGroup => (
            <div
              {...headerGroup.getHeaderGroupProps({
                style: { paddingRight: '15px' }
              })}
              className="tr"
            >
              {headerGroup.headers.map(column => {
                return (
                  <div className="th" {...column.getHeaderProps(headerProps)}>
                    {column.render('Header')}
                    <span>
                      {column.isSorted
                        ? column.isSortedDesc
                          ? ' ▼'
                          : ' ▲'
                        : ' '}
                    </span>
                  </div>
                )
              })}
            </div>
          ))}
        </div>
        <div className="tbody">
          {rows.map(row => {
            prepareRow(row)
            return (
              <div {...row.getRowProps()} className="tr styled-table-row">
                {row.cells.map(cell => (
                  <div {...cell.getCellProps(cellProps)} className="td">
                    {cell.render('Cell')}
                  </div>
                ))}
              </div>
            )
          })}
        </div>
      </div>
    </div>
  )
}
