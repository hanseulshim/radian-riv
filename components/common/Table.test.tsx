import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Table from './Table'

describe('Table', () => {
  let component
  beforeAll(() => {
    const props = {
      columns: [],
      data: [],
      fetchData: () => {}
    }
    component = shallow(<Table {...props} />)
  })
  it('renders table container', () => {
    expect(component.find('.table-container')).toHaveLength(1)
  })
  it('renders table', () => {
    expect(component.find('.table')).toHaveLength(1)
    expect(component.find('.styled-table-row')).toHaveLength(0)
  })
  it('should match snapshot after clicking image', () => {
    expect(toJson(component)).toMatchSnapshot()
  })
})
