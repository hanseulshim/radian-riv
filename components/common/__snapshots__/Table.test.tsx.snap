// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot after clicking image 1`] = `
<div
  className="table-container"
  style={
    Object {
      "width": undefined,
    }
  }
>
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
    />
    <div
      className="tbody"
    />
  </div>
</div>
`;
